File: 36. Valid Sudoku/Approach 2 - Hash Set/Valid Sudoku.py
class Solution:
    def isValidSudoku(self, board: List[List[str]]) -> bool:
        N = 9
        rowSet = [set() for _ in range(N)]
        colSet = [set() for _ in range(N)]
        squareSet = [set() for _ in range(N)]

        for r in range(N):
            for c in range(N):
                if board[r][c] == ".": continue
                sr, sc = r // 3, c // 3
                sPos = sr * 3 + sc
                if board[r][c] in rowSet[r] or board[r][c] in colSet[c] or board[r][c] in squareSet[sPos]:
                    return False
                rowSet[r].add(board[r][c])
                colSet[c].add(board[r][c])
                squareSet[sPos].add(board[r][c])

        return True

File: 36. Valid Sudoku/Approach 2 - Hash Set/Valid Sudoku.java
class Solution {
    public boolean isValidSudoku(char[][] board) {
        int N = 9;
        HashSet<Character>[] rowSet = new HashSet[N];
        HashSet<Character>[] colSet = new HashSet[N];
        HashSet<Character>[] squareSet = new HashSet[N];
        for (int r = 0; r < N; r++) {
            rowSet[r] = new HashSet<>();
            colSet[r] = new HashSet<>();
            squareSet[r] = new HashSet<>();
        }
        for (int r = 0; r < N; r++) {
            for (int c = 0; c < N; c++) {
                char val = board[r][c];
                if (val == '.') continue;
                int sPos = (r / 3) * 3 + c / 3;
                if (rowSet[r].contains(val) || colSet[c].contains(val) || squareSet[sPos].contains(val))
                    return false;
                rowSet[r].add(val);
                colSet[c].add(val);
                squareSet[sPos].add(val);
            }
        }
        return true;
    }
}